File: bindings/python/examples/record_store.py
"""Store typed dictation notes, page them, and run a filtered search."""

from __future__ import annotations

import tempfile
from pathlib import Path

import numpy as np

import zeppelin_embed as ze


def main() -> int:
    schema = ze.NamespaceSpec(
        attributes=(
            ze.AttributeDefinition(1, "priority", ze.AttributeType.U64),
            ze.AttributeDefinition(2, "reviewed", ze.AttributeType.BOOL),
            ze.AttributeDefinition(3, "category", ze.AttributeType.DICTIONARY_STRING),
            ze.AttributeDefinition(4, "project", ze.AttributeType.RAW_STRING, nullable=True),
        ),
        vector_space=ze.VectorSpace(2, normalization=ze.Normalization.UNIT_L2),
    )

    # Each namespace lives at root/name on disk. The temporary root keeps the
    # example self-cleaning; use an application path to reopen the data later.
    with tempfile.TemporaryDirectory() as scratch:
        root = Path(scratch)
        with ze.open_namespace(root, "notes", schema) as notes:
            documents = (
                ze.StoredDocument(
                    101,
                    timestamp=100,
                    vector=np.asarray([1.0, 0.0], dtype=np.float32),
                    text="Plan the product launch",
                    attributes=(
                        ze.AttributeValue(1, ze.AttributeType.U64, 2),
                        ze.AttributeValue(2, ze.AttributeType.BOOL, True),
                        ze.AttributeValue(3, ze.AttributeType.DICTIONARY_STRING, "work"),
                        ze.AttributeValue(4, ze.AttributeType.RAW_STRING, "zeppelin"),
                    ),
                ),
                ze.StoredDocument(
                    102,
                    timestamp=300,
                    vector=np.asarray([0.0, 1.0], dtype=np.float32),
                    text="Buy oat milk",
                    attributes=(
                        ze.AttributeValue(1, ze.AttributeType.U64, 1),
                        ze.AttributeValue(2, ze.AttributeType.BOOL, False),
                        ze.AttributeValue(3, ze.AttributeType.DICTIONARY_STRING, "personal"),
                        ze.AttributeValue(4, ze.AttributeType.RAW_STRING, None),
                    ),
                ),
                ze.StoredDocument(
                    103,
                    timestamp=200,
                    vector=np.asarray([0.8, 0.6], dtype=np.float32),
                    text="Review search benchmarks",
                    attributes=(
                        ze.AttributeValue(1, ze.AttributeType.U64, 3),
                        ze.AttributeValue(2, ze.AttributeType.BOOL, True),
                        ze.AttributeValue(3, ze.AttributeType.DICTIONARY_STRING, "work"),
                        ze.AttributeValue(4, ze.AttributeType.RAW_STRING, "zeppelin"),
                    ),
                ),
                ze.StoredDocument(
                    104,
                    timestamp=400,
                    vector=np.asarray([0.6, 0.8], dtype=np.float32),
                    text="Book dentist appointment",
                    attributes=(
                        ze.AttributeValue(1, ze.AttributeType.U64, 2),
                        ze.AttributeValue(2, ze.AttributeType.BOOL, False),
                        ze.AttributeValue(3, ze.AttributeType.DICTIONARY_STRING, "personal"),
                        ze.AttributeValue(4, ze.AttributeType.RAW_STRING, None),
                    ),
                ),
            )
            mutation = notes.upsert(documents)
            print(f"upserted 4 notes at generation {mutation.generation}")

            # get preserves caller order and returns None for a missing id.
            lookup = notes.get([101, 999])
            found, missing = lookup.documents
            if found is None:
                raise RuntimeError("note 101 unexpectedly missing")
            print(f'get 101: "{found.text}"')
            if missing is None:
                print(f"get 999: missing ({lookup.missing_count} missing)")

            # The binding flattens this structured AND into the C ABI's
            # contiguous child-node range.
            selected = ze.Filter.and_(
                ze.Filter.eq("category", "work"),
                ze.Filter.range_("priority", gte=2, lte=3),
            )
            cursor = None
            scanned = 0
            page_number = 1
            while True:
                page = notes.scan(
                    cursor=cursor,
                    limit=1,
                    order="timestamp_ascending",
                    filter=selected,
                )
                for note in page.documents:
                    print(
                        f'scan page {page_number}: note {note.doc_id} '
                        f'at {note.timestamp}: "{note.text}"'
                    )
                scanned += len(page.documents)
                cursor = page.cursor
                if cursor is None:
                    break
                page_number += 1

            count = notes.count(filter=selected)
            print(f"count: {count.count} matching notes (scan found {scanned})")

            matches = notes.search(
                np.asarray([1.0, 0.0], dtype=np.float32),
                k=2,
                tier=ze.Tier.EXACT,
                filter=selected,
            )
            for rank, hit in enumerate(matches.hits, start=1):
                print(f"search {rank}: note {hit.doc_id}, score {hit.score:.3f}")

        # Omitting a vector space creates a plain record store. Vector search
        # on this namespace is rejected with ZE_ERR_NO_VECTOR_SPACE.
        with ze.open_namespace(root, "inbox", ze.NamespaceSpec()) as inbox:
            inbox.upsert([ze.StoredDocument(201, timestamp=500, text="Call Alice")])
            record = inbox.scan(limit=10, order="timestamp_ascending").documents[0]
            print(f'record-only scan: note {record.doc_id}: "{record.text}"')

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bindings/node/examples/record-store.ts
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import zeppelinEmbed from '../index.js';
import type {
  Filter,
  NamespaceSpec,
  ScanCursor,
  UpsertDocument,
} from '../index.js';

const { openNamespace } = zeppelinEmbed;

const spec: NamespaceSpec = {
  attributes: [
    { id: 1, name: 'priority', type: 'u64' },
    { id: 2, name: 'reviewed', type: 'bool' },
    { id: 3, name: 'category', type: 'dictionaryString' },
    { id: 4, name: 'project', type: 'rawString', nullable: true },
  ],
  vectorSpace: { dimensions: 2, normalization: 'unitL2' },
};

// Each namespace lives at root/name on disk. The temporary root keeps the
// example self-cleaning; use an application path to reopen the data later.
const root = mkdtempSync(join(tmpdir(), 'zeppelin-node-record-store-'));
const notes = openNamespace(root, 'notes', spec);

try {
  const documents: UpsertDocument[] = [
    {
      id: 101n,
      timestamp: 100n,
      vector: new Float32Array([1, 0]),
      text: 'Plan the product launch',
      attributes: [
        { id: 1, type: 'u64', value: 2n },
        { id: 2, type: 'bool', value: true },
        { id: 3, type: 'string', value: 'work' },
        { id: 4, type: 'string', value: 'zeppelin' },
      ],
    },
    {
      id: 102n,
      timestamp: 300n,
      vector: new Float32Array([0, 1]),
      text: 'Buy oat milk',
      attributes: [
        { id: 1, type: 'u64', value: 1n },
        { id: 2, type: 'bool', value: false },
        { id: 3, type: 'string', value: 'personal' },
        { id: 4, type: 'null', value: null },
      ],
    },
    {
      id: 103n,
      timestamp: 200n,
      vector: new Float32Array([0.8, 0.6]),
      text: 'Review search benchmarks',
      attributes: [
        { id: 1, type: 'u64', value: 3n },
        { id: 2, type: 'bool', value: true },
        { id: 3, type: 'string', value: 'work' },
        { id: 4, type: 'string', value: 'zeppelin' },
      ],
    },
    {
      id: 104n,
      timestamp: 400n,
      vector: new Float32Array([0.6, 0.8]),
      text: 'Book dentist appointment',
      attributes: [
        { id: 1, type: 'u64', value: 2n },
        { id: 2, type: 'bool', value: false },
        { id: 3, type: 'string', value: 'personal' },
        { id: 4, type: 'null', value: null },
      ],
    },
  ];
  const mutation = notes.upsert(documents);
  console.log(`upserted 4 notes at generation ${mutation.generation}`);

  // get preserves caller order and returns null for a missing id.
  const lookup = notes.get([101n, 999n]);
  const [found, missing] = lookup.documents;
  if (found === null || found === undefined) {
    throw new Error('note 101 unexpectedly missing');
  }
  console.log(`get 101: "${found.text}"`);
  if (missing === null) {
    console.log(`get 999: missing (${lookup.missingCount} missing)`);
  }

  // Logical children become one consecutive range in the C ABI's flat node
  // array; callers keep the readable nested form.
  const selected: Filter = {
    op: 'and',
    children: [
      {
        op: 'eq',
        attributeId: 3,
        values: [{ id: 3, type: 'string', value: 'work' }],
      },
      {
        op: 'range',
        attributeId: 1,
        lower: { id: 1, type: 'u64', value: 2n },
        lowerInclusive: true,
        upper: { id: 1, type: 'u64', value: 3n },
        upperInclusive: true,
      },
    ],
  };
  let cursor: ScanCursor | undefined;
  let scanned = 0;
  let pageNumber = 1;
  do {
    const page = notes.scan({
      cursor,
      limit: 1,
      order: 'timestampAscending',
      filter: selected,
    });
    for (const note of page.documents) {
      console.log(
        `scan page ${pageNumber}: note ${note.id} at ${note.timestamp}: "${note.text}"`,
      );
    }
    scanned += page.documents.length;
    cursor = page.cursor ?? undefined;
    pageNumber += 1;
  } while (cursor !== undefined);

  const count = notes.count({ filter: selected });
  console.log(`count: ${count.count} matching notes (scan found ${scanned})`);

  const matches = notes.searchFiltered(
    new Float32Array([1, 0]),
    selected,
    { k: 2, tier: 'exact' },
  );
  for (const [index, hit] of matches.entries()) {
    console.log(`search ${index + 1}: note ${hit.id}, score ${hit.score.toFixed(3)}`);
  }
} finally {
  notes.close();
}

// Omitting a vector space creates a plain record store. Vector search on this
// namespace is rejected with ZE_ERR_NO_VECTOR_SPACE.
const inbox = openNamespace(root, 'inbox', { attributes: [] });
try {
  inbox.upsert([{ id: 201n, timestamp: 500n, text: 'Call Alice' }]);
  const record = inbox.scan({ limit: 10, order: 'timestampAscending' }).documents[0];
  if (record === undefined) {
    throw new Error('record-only scan unexpectedly empty');
  }
  console.log(`record-only scan: note ${record.id}: "${record.text}"`);
} finally {
  inbox.close();
  rmSync(root, { force: true, recursive: true });
}
